src/main/java/com/hierynomus/smbj/lock/CrossThreadLock.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C)2016 - SMBJ Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hierynomus.smbj.lock;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * A simple, non-reentrant lock mechanism, which unlike {@link ReentrantLock} allows a thread to release a lock acquired
+ * by a different thread.
+ */
+public class CrossThreadLock {
```

Comment:
Not sure what you mean about being non-safe. It passed some stress-testing using the test class, with larger parameters than the default.
Didn't find an OOB lock implementation that would work across threads. Actually I was kind of surprised that ReentrantLock doesn't. Maybe there is one in third-party code, I preferred not to introduce new dependencies.